### PR TITLE
add support for hillshading on OSM

### PR DESCRIPTION
**Pull request by ztNFny**
Implements support for hillshading in OSM maps.
UnifiedMap is not covered by this PR.

DEM files have to be obtained manually. Good sources I know are https://sonny.4lima.de or http://viewfinderpanoramas.org/dem3.html - this should be added to FAQ before a release.

fixes https://github.com/cgeo/cgeo/issues/12210

### CODE REVIEW

**Comment by ztNFny** [comment]
There's 3 Codacy errors I have no idea what to do about:
```
main/src/cgeo/geocaching/maps/mapsforge/v6/layers/HillShadingLayer.java#L15
The utility class name 'HillShadingLayer' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'

main/src/cgeo/geocaching/maps/mapsforge/v6/layers/HillShadingLayer.java#L15
Utility classes should not have a public or default constructor.
```

Why is it marked a "utility class"? Naming matches the other classes in that package.

```
main/src/cgeo/geocaching/maps/mapsforge/v6/layers/RendererLayer.java#L4
Unused import - org.mapsforge.map.android.graphics.AndroidGraphicFactory.
```

The import is used though, as "AndroidGraphicFactory.INSTANCE"

**Comment by fm-sys** [comment]
> ```
> main/src/cgeo/geocaching/maps/mapsforge/v6/layers/HillShadingLayer.java#L15
> Utility classes should not have a public or default constructor.
> ```
> 
> 
>     
>   
> 
> Why is it marked a "utility class"?

It's marked as an utility class cause all methods inside inside are static. Therefore, no instances need to be created and the automatically generated default constructor is useless.

Please add something like this to your code:

```
private HillShadingLayer() {
    // Do not instantiate, utility class
} 
```

For the naming, IMHO `HillShadingHelper` would work as well (especially as it's technically not returning a layer instance anyway). But we can also leave it as is.

>     Unused import - org.mapsforge.map.android.graphics.AndroidGraphicFactory.
>
> The import is used though, as "AndroidGraphicFactory.INSTANCE"

I don't see where `AndroidGraphicFactory.INSTANCE` should be used...?

**Comment by ztNFny** [comment]
> I don't see where `AndroidGraphicFactory.INSTANCE` should be used...?

new MemoryCachingHgtReaderTileSource(shadingFolder, new SimpleShadingAlgorithm(Settings.getMapShadingLinearity(), Settings.getMapShadingScale()), **AndroidGraphicFactory.INSTANCE**);

**Comment by ztNFny** [comment]
> This line is in `HillShadingLayer` while codacy is complaining about an unused import in `RendererLayer`
> ;-)

ah fu... 

Adding the private constructor didn't get rid of the "The utility class name 'HillShadingLayer' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'" error?!

**Comment by eddiemuc** [comment]
> > This line is in `HillShadingLayer` while codacy is complaining about an unused import in `RendererLayer`
> > ;-)
> 
> ah fu...
> 
> Adding the private constructor didn't get rid of the "The utility class name 'HillShadingLayer' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'" error?!

Rename it to e.g. `HillShadingLayerHelper` or `HillShadingLayerUtils` will solve the error.

File: main/res/values/preference_keys.xml [hunk 1]
```diff
@@ -190,6 +190,7 @@
     <string translatable="false" name="pref_persistablefolder_offlinemaps">persistablefolder_offlinemaps</string>
     <string translatable="false" name="pref_fakekey_info_offline_mapthemes">fakekey_info_offline_mapthemes</string>
     <string translatable="false" name="pref_persistablefolder_offlinemapthemes">persistablefolder_offlinemapthemes</string>
+    <string translatable="false" name="pref_persistablefolder_offlinemapshading">pref_persistablefolder_offlinemapshading</string>
```
Comment (moving-bits) [inline main/res/values/preference_keys.xml:193]: Please do not use "pref_" prefix for the preference variable value (only for preference variable name).

File: main/res/values/preference_keys.xml [hunk 2]
```diff
@@ -396,6 +397,9 @@
         <item>@string/pref_value_theme_dark</item>
     </string-array>
 
+    <string translatable="false" name="pref_mapShadingLinearity">pref_mapShadingLinearity</string>
+    <string translatable="false" name="pref_mapShadingScale">pref_mapShadingScale</string>
+
```
Comment (moving-bits) [inline main/res/values/preference_keys.xml:403]: Same as above (do not use "pref_" prefix for preference variable value)

File: main/res/values/preference_keys.xml [hunk 3]
```diff
@@ -396,6 +397,9 @@
         <item>@string/pref_value_theme_dark</item>
     </string-array>
 
+    <string translatable="false" name="pref_mapShadingLinearity">pref_mapShadingLinearity</string>
```
Comment (moving-bits) [inline main/res/values/preference_keys.xml:400]: Same as above (do not use "pref_" prefix for preference variable value)

File: main/res/xml/preferences_map.xml [hunk 4]
```diff
@@ -70,6 +70,29 @@
             android:summary="%s"
             android:title="@string/settings_title_map_rotation"
             app:iconSpaceReserved="false" />
+        <Preference
```
Comment (moving-bits) [inline main/res/xml/preferences_map.xml:73]: Maybe shading could get its own preference category, putting some of the text in "init_mapshading_folder_text" into the category's description?

File: main/src/cgeo/geocaching/storage/PersistableFolder.java [hunk 5]
```diff
@@ -40,6 +40,8 @@ public enum PersistableFolder {
      */
     //legacy setting: "renderthemepath", a pure file path is stored
     OFFLINE_MAP_THEMES(R.string.pref_persistablefolder_offlinemapthemes, R.string.persistablefolder_offline_maps_themes, Folder.fromPersistableFolder(OFFLINE_MAPS, "_themes")),
+    /** Offline Maps: optional folder for map shading files (dem) */
+    OFFLINE_MAP_SHADING(R.string.pref_persistablefolder_offlinemapshading, R.string.pref_persistablefolder_offlinemapshading, Folder.fromPersistableFolder(OFFLINE_MAPS, "_hgt")),
```
Comment (moving-bits) [inline main/src/cgeo/geocaching/storage/PersistableFolder.java:44]: second parameter should be the preference's title, not it's preference key name - current implementation leads to a strange dialog content (title and text) when setting the folder

**Review by moving-bits — CHANGES_REQUESTED** [review]
Thanks for diving into hillshading and SAF. Implementation seems to work generally, although I got some NPE in individual routes - not sure if it's related to this PR, though.
Some texts may need some fine-tuning for more clarity, but this can be done later.
Otherwise I have some minor findings, see review comments.

File: main/res/xml/preferences_map.xml (hunk 4 above)
Comment (ztNFny) [reply main/res/xml/preferences_map.xml:73]: Not sure about this one. Map themes, which is kinda similar, doesn't have its own section either.

I'm also not sure whether pref_mapShadingLinearity and pref_mapShadingScale really need to be user-configurable settings. I've tried to find reasonable defaults for those.
Maybe we could get some feedback during beta whether the defaults are good enough and if they are remove those settings (from GUI at least).

**Comment by ztNFny** [comment]
retest this please

File: main/res/xml/preferences_map.xml (hunk 4 above)
Comment (moving-bits) [reply main/res/xml/preferences_map.xml:73]: Yes, could be a way to go.

Comment (ztNFny) [reply main/res/xml/preferences_map.xml:73]: Let's merge then - your Review is still blocking it somehow

**Comment by fm-sys** [comment]
retest this please

**Comment by moving-bits** [comment]
@mucek4 
It's the "Could not create ADB Bridge." error again, can you restart the executors, please?

File: main/res/xml/preferences_map.xml (hunk 4 above)
Comment (moving-bits) [reply main/res/xml/preferences_map.xml:73]: > Let's merge then - your Review is still blocking it somehow

Open reviews or reviews requesting changes do not block the build process and do not lead to failing builds.
It's an error in our CI environment, sometimes the server fails with "Cannot create ADB bridge", so it cannot run the tests - which leads to a failed / red build.

Let's wait for @mucek4 to fix this, and retest the PR then.

**Review by moving-bits — APPROVED** [review]
LGTM, thanks for fixing the issues.

**Comment by moving-bits** [comment]
retest this please

**Pull request merged**